File: mira-api/src/ai/mocks/mock-outfit-analysis.provider.ts
import { Injectable } from '@nestjs/common';
import {
  MIRA_OCCASION_LABELS,
  MiraOccasion,
  occasionIndex,
} from '../contracts/mira-occasion';
import { OutfitAnalysisResult } from '../contracts/outfit-analysis-result.interface';
import { OutfitStyleMetrics } from '../contracts/outfit-style-metrics.interface';
import { OutfitAnalysisProvider } from '../providers/outfit-analysis.provider';
import { delay, nextInt, seedFromImageBytes } from '../utils/image-seed';

/**
 * Phase 0 LEGACY — mock outfit provider for tests/dev only.
 * Blocked from serving production results via OutfitAnalysisService guard.
 * Canonical path: Vision Platform /ai/vision/outfit/analyze
 */
@Injectable()
export class MockOutfitAnalysisProvider implements OutfitAnalysisProvider {
  private readonly garments = [
    { ar: 'فستان', en: 'Dress' },
    { ar: 'عباءة', en: 'Abaya' },
    { ar: 'بدلة', en: 'Suit' },
    { ar: 'تنورة وبلوزة', en: 'Skirt & Blouse' },
    { ar: 'جumpsuit', en: 'Jumpsuit' },
  ];

  private readonly styles = [
    { ar: 'أنيق', en: 'Elegant' },
    { ar: 'كلاسيكي', en: 'Classic' },
    { ar: 'عصري', en: 'Modern' },
    { ar: 'راقي', en: 'Refined' },
    { ar: 'بسيط', en: 'Minimal' },
  ];

  private readonly colorPalettes = [
    ['زيتوني', 'ذهبي'],
    ['وردي', 'بيج'],
    ['أسود', 'فضي'],
    ['أزرق ملكي', 'كريمي'],
    ['نبيتي', 'ذهبي'],
    ['رمادي', 'أبيض'],
    ['مرجاني', 'بيج'],
  ];

  async analyze(
    imageBytes: Buffer,
    occasion: MiraOccasion,
  ): Promise<OutfitAnalysisResult> {
    await delay(850);

    const seed = seedFromImageBytes(imageBytes) ^ occasionIndex(occasion) * 997;
    const rng = { seed };

    const garment = this.garments[seed % this.garments.length];
    const style = this.styles[Math.floor(seed / 3) % this.styles.length];
    const palette = this.colorPalettes[Math.floor(seed / 5) % this.colorPalettes.length];
    const styleMetrics = this.generateMetrics(seed, occasion, rng);
    const provisionalScore = this.provisionalScore(styleMetrics);

    const labels = MIRA_OCCASION_LABELS[occasion];
    const suitability = this.occasionSuitability(occasion, provisionalScore, labels);

    return {
      compatibilityScore: provisionalScore,
      dominantColors: [...palette],
      garmentTypeAr: garment.ar,
      garmentTypeEn: garment.en,
      styleCategoryAr: style.ar,
      styleCategoryEn: style.en,
      occasionSuitabilityAr: suitability.ar,
      occasionSuitabilityEn: suitability.en,
      alternativeColorsAr: this.alternativeColorsAr(occasion),
      alternativeColorsEn: this.alternativeColorsEn(occasion),
      occasion,
      styleMetrics,
    };
  }

  private generateMetrics(
    seed: number,
    occasion: MiraOccasion,
    rng: { seed: number },
  ): OutfitStyleMetrics {
    const base = 44 + nextInt(rng, 38);
    const occasionBias = this.occasionBias(occasion, seed);

    const colorHarmony = clamp(base + nextInt(rng, 14) - 4, 0, 100);
    const occasionFit = clamp(base + occasionBias + nextInt(rng, 12) - 6, 0, 100);
    const styleCoherence = clamp(base + nextInt(rng, 10) - 3, 0, 100);
    const silhouetteBalance = clamp(base + nextInt(rng, 12) - 5, 0, 100);
    const polish = clamp(base + nextInt(rng, 10) - 6, 0, 100);

    return {
      colorHarmony,
      occasionFit,
      styleCoherence,
      silhouetteBalance,
      polish,
      colorClashSeverity: clamp(100 - colorHarmony + nextInt(rng, 16), 0, 100),
      occasionMismatchSeverity: clamp(100 - occasionFit + nextInt(rng, 14), 0, 100),
      tonalImbalanceSeverity: clamp(
        100 - Math.round((colorHarmony + styleCoherence) / 2) + nextInt(rng, 12),
        0,
        100,
      ),
      accessoryOverloadSeverity: clamp(18 + nextInt(rng, 40), 0, 100),
      formalityGapSeverity: clamp(100 - occasionFit + nextInt(rng, 10), 0, 100),
    };
  }

  private occasionBias(occasion: MiraOccasion, seed: number): number {
    switch (occasion) {
      case MiraOccasion.Wedding:
      case MiraOccasion.Interview:
        return seed % 2 === 0 ? 6 : -8;
      case MiraOccasion.Casual:
      case MiraOccasion.University:
        return 4;
      default:
        return 0;
    }
  }

  private provisionalScore(metrics: OutfitStyleMetrics): number {
    const positive =
      metrics.colorHarmony * 0.28 +
      metrics.occasionFit * 0.26 +
      metrics.styleCoherence * 0.18 +
      metrics.silhouetteBalance * 0.14 +
      metrics.polish * 0.14;
    return clamp(Math.round(positive * 0.6 + 22), 0, 100);
  }

  private occasionSuitability(
    occasion: MiraOccasion,
    score: number,
    labels: { ar: string; en: string },
  ): { ar: string; en: string } {
    const level =
      score >= 86
        ? { ar: 'ممتاز', en: 'Excellent' }
        : score >= 74
          ? { ar: 'مناسب جدًا', en: 'Very suitable' }
          : score >= 62
            ? { ar: 'مناسب', en: 'Suitable' }
            : score >= 48
              ? { ar: 'يحتاج تحسين', en: 'Needs improvement' }
              : { ar: 'غير مناسب حالياً', en: 'Not suitable yet' };

    return {
      ar: `${level.ar} لمناسبة ${labels.ar}`,
      en: `${level.en} for ${labels.en}`,
    };
  }

  private alternativeColorsAr(occasion: MiraOccasion): string[] {
    switch (occasion) {
      case MiraOccasion.Wedding:
      case MiraOccasion.Eid:
        return ['ذهبي', 'شامبين', 'عنابي', 'وردي-soft'];
      case MiraOccasion.Work:
      case MiraOccasion.Interview:
        return ['كحلي', 'بيج', 'رمادي فاتح', 'أبيض كريمي'];
      case MiraOccasion.Evening:
        return ['أسود', 'ياقوتي', 'فضي', 'زيتي داكن'];
      default:
        return ['تركواز', 'مرجاني', 'لافندر', 'بيج'];
    }
  }

  private alternativeColorsEn(occasion: MiraOccasion): string[] {
    switch (occasion) {
      case MiraOccasion.Wedding:
      case MiraOccasion.Eid:
        return ['Gold', 'Champagne', 'Burgundy', 'Soft Pink'];
      case MiraOccasion.Work:
      case MiraOccasion.Interview:
        return ['Navy', 'Beige', 'Light Grey', 'Cream White'];
      case MiraOccasion.Evening:
        return ['Black', 'Ruby', 'Silver', 'Dark Olive'];
      default:
        return ['Turquoise', 'Coral', 'Lavender', 'Beige'];
    }
  }
}

function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

File: lib/core/ai/mocks/mock_outfit_analysis_provider.dart
import 'dart:math' as math;

import '../models/mira_occasion.dart';
import '../models/outfit_analysis_result.dart';
import '../providers/outfit_analysis_provider.dart';
import '../utils/image_seed.dart';
import '../../../features/outfit_analysis/domain/entities/outfit_style_metrics.dart';

/// Deterministic mock — simulates FASHN.ai until API key is configured.
class MockOutfitAnalysisProvider implements OutfitAnalysisProvider {
  static const _garments = [
    (ar: 'فستان', en: 'Dress'),
    (ar: 'عباءة', en: 'Abaya'),
    (ar: 'بدلة', en: 'Suit'),
    (ar: 'تنورة وبلوزة', en: 'Skirt & Blouse'),
  ];

  static const _styles = [
    (ar: 'أنيق', en: 'Elegant'),
    (ar: 'كلاسيكي', en: 'Classic'),
    (ar: 'عصري', en: 'Modern'),
    (ar: 'راقي', en: 'Refined'),
  ];

  static const _colorPalettes = [
    ['زيتوني', 'ذهبي'],
    ['وردي', 'بيج'],
    ['أسود', 'فضي'],
    ['أزرق ملكي', 'كريمي'],
    ['نبيتي', 'ذهبي'],
  ];

  @override
  Future<OutfitAnalysisResult> analyze({
    required List<int> imageBytes,
    required MiraOccasion occasion,
  }) async {
    await Future<void>.delayed(const Duration(milliseconds: 850));

    final seed = seedFromImageBytes(imageBytes) ^ occasion.index * 997;
    final rng = math.Random(seed);

    final garment = _garments[seed % _garments.length];
    final style = _styles[(seed ~/ 3) % _styles.length];
    final palette = _colorPalettes[(seed ~/ 5) % _colorPalettes.length];

    final baseHealth = 52 + rng.nextInt(38);
    final styleMetrics = _buildMetrics(baseHealth, rng, occasion);
    final compatibilityScore = baseHealth.toDouble();
    final suitability = _occasionSuitability(occasion, compatibilityScore);

    return OutfitAnalysisResult(
      compatibilityScore: compatibilityScore,
      dominantColors: List<String>.from(palette),
      garmentTypeAr: garment.ar,
      garmentTypeEn: garment.en,
      styleCategoryAr: style.ar,
      styleCategoryEn: style.en,
      occasionSuitabilityAr: suitability.ar,
      occasionSuitabilityEn: suitability.en,
      alternativeColorsAr: _alternativeColorsAr(occasion),
      alternativeColorsEn: _alternativeColorsEn(occasion),
      occasion: occasion,
      styleMetrics: styleMetrics,
    );
  }

  static OutfitStyleMetrics _buildMetrics(
    int baseHealth,
    math.Random rng,
    MiraOccasion occasion,
  ) {
    int jitter(int center, int spread) =>
        (center + rng.nextInt(spread * 2 + 1) - spread).clamp(0, 100);

    var occasionFit = jitter(baseHealth, 8);
    var formalityGap = (100 - baseHealth + rng.nextInt(12)).clamp(0, 100);

    if (occasion == MiraOccasion.interview || occasion == MiraOccasion.work) {
      occasionFit = (occasionFit - 6).clamp(0, 100);
      formalityGap = (formalityGap + 8).clamp(0, 100);
    }

    return OutfitStyleMetrics(
      colorHarmony: jitter(baseHealth, 6),
      occasionFit: occasionFit,
      styleCoherence: jitter(baseHealth - 2, 7),
      silhouetteBalance: jitter(baseHealth - 3, 6),
      polish: jitter(baseHealth - 5, 8),
      colorClashSeverity: (100 - baseHealth + rng.nextInt(10)).clamp(0, 100),
      occasionMismatchSeverity: (100 - occasionFit + rng.nextInt(8)).clamp(0, 100),
      tonalImbalanceSeverity: (100 - baseHealth + rng.nextInt(14)).clamp(0, 100),
      accessoryOverloadSeverity: jitter(100 - baseHealth, 12),
      formalityGapSeverity: formalityGap,
    );
  }

  static ({String ar, String en}) _occasionSuitability(
    MiraOccasion occasion,
    double score,
  ) {
    final level = score >= 86
        ? (ar: 'ممتاز', en: 'Excellent')
        : score >= 74
            ? (ar: 'مناسب جدًا', en: 'Very suitable')
            : score >= 62
                ? (ar: 'مناسب', en: 'Suitable')
                : (ar: 'يحتاج تحسين', en: 'Needs improvement');

    return (
      ar: '${level.ar} لمناسبة ${occasion.labelAr}',
      en: '${level.en} for ${occasion.labelEn}',
    );
  }

  static List<String> _alternativeColorsAr(MiraOccasion occasion) {
    switch (occasion) {
      case MiraOccasion.wedding:
      case MiraOccasion.eid:
        return ['ذهبي', 'شامبين', 'عنابي'];
      case MiraOccasion.work:
      case MiraOccasion.interview:
        return ['كحلي', 'بيج', 'رمادي فاتح'];
      case MiraOccasion.evening:
        return ['أسود', 'ياقوتي', 'فضي'];
      default:
        return ['تركواز', 'مرجاني', 'لافندر'];
    }
  }

  static List<String> _alternativeColorsEn(MiraOccasion occasion) {
    switch (occasion) {
      case MiraOccasion.wedding:
      case MiraOccasion.eid:
        return ['Gold', 'Champagne', 'Burgundy'];
      case MiraOccasion.work:
      case MiraOccasion.interview:
        return ['Navy', 'Beige', 'Light Grey'];
      case MiraOccasion.evening:
        return ['Black', 'Ruby', 'Silver'];
      default:
        return ['Turquoise', 'Coral', 'Lavender'];
    }
  }
}
